### PR TITLE
kernel: Implement k_sleep for Single Thread

### PR DESCRIPTION
**Pull request by nandojve**
The current z_tick_sleep return directly when building kernel for Single Thread model. This reorganize the code and implement a spin lock to wait pass the required ticks.

### CODE REVIEW

File: kernel/sched.c
Comment (otavio) [inline kernel/sched.c:1476]: It'd be good to have a small comment in the code explaining why this is need here.

**Comment by hongshui3000** [comment]
I'm also curious, we always k_wait in single-threaded applications, why do we need k_sleep?

**Comment by nandojve** [comment]
> I'm also curious, we always k_wait in single-threaded applications, why do we need k_sleep?

The intention is not to provide more kernel functionality to Single Thread applications instead is to make drivers and subsystems available to it. For instance, I need RTC driver from ST and it uses a k_msleep
https://github.com/zephyrproject-rtos/zephyr/blob/b272760f062b36fe6ce5d14c7a3d9065d40a570a/drivers/rtc/rtc_ll_stm32.c#L84

The intention is make it compatible with the whole system at all. In this case the driver was wrote to multi-thread but it can be used in Single Thread applications.

> I think we should just remove #ifdef CONFIG_MULTITHREADING instead of changing k_sleep so that it can be used in single thread

If I understood you correctly I'm not sure if this is possible.

> sched.c is only compiled when using multiple threads. The limitation of CONFIG_MULTITHREADING seems redundant.

I originally opened at #63987 and than I realize that a spin look to wait the ticks is needed to make RTC works.

**Comment by hongshui3000** [comment]
> > I'm also curious, we always k_wait in single-threaded applications, why do we need k_sleep?
> 
> The intention is not to provide more kernel functionality to Single Thread applications instead is to make drivers and subsystems available to it. For instance, I need RTC driver from ST and it uses a k_msleep
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/b272760f062b36fe6ce5d14c7a3d9065d40a570a/drivers/rtc/rtc_ll_stm32.c#L84
> 
> The intention is make it compatible with the whole system at all. In this case the driver was wrote to multi-thread but it can be used in Single Thread applications.


 #ifdef CONFIG_MULTITHREADING
bool success = WAIT_FOR(LL_RTC_IsActiveFlag_INIT(RTC), RTC_TIMEOUT, k_msleep(1)); 
#else
bool success = WAIT_FOR(LL_RTC_IsActiveFlag_INIT(RTC), RTC_TIMEOUT, k_wait(1000)); 
#endif
Wouldn't it be better to do this? According to the documentation of zephyr rtos, k_msleep will not actually appear in multi-threaded applications. You may need to change the document according to your change method

**Comment by nandojve** [comment]
> Wouldn't it be better to do this?

IMHO a good idea is make a small patch on a single point that do not affect anything and enable all subsystem/drivers that use the primitive. It's far better than start to adapt drivers/subsystems. It is perfect reasonable if there is intention to make code compatible.

Do you see issues to make `k_sleep` behave exactly like `k_wait` for single thread ?

> You may need to change the document according to your change method.

Sure, I'll do it.

**Comment by hongshui3000** [comment]
> > Wouldn't it be better to do this?
> 
> IMHO a good idea is make a small patch on a single point that do not affect anything and enable all subsystem/drivers that use the primitive. It's far better than start to adapt drivers/subsystems. It is perfect reasonable if there is intention to make code compatible.
> 
> Do you see issues to make `k_sleep` behave exactly like `k_wait` for single thread ?
> 
> > You may need to change the document according to your change method.
> 
> Sure, I'll do it.

I am not opposed to this change. One bad thing I think is that there are many functions in the sched.c file that cannot be used in a single thread. After the current change, you cannot get a warning during the compilation phase. This is where I think there's a problem

**Comment by hongshui3000** [comment]
Would it be better to combine k_sleep and k_busy_wait into one file. Make them usable from a single thread. It is obvious that zephyr's scheduler(sched.c) cannot be used in a single thread

**Comment by peter-mitsis** [comment]
I don't think that moving sched.c to always be built is the right direction.

Two ideas come to mind that build upon each other--perhaps they are not any good, but I will toss them out for evaluation nonetheless.

The first is to have a non-multithreading definition either in its own file, or in _busy_wait.c_.

The second is ... can we build on this alternate _k_sleep()_ to leverage k_timer in some way? This would require some exploration/experimentation, but the gist of this idea is to set up a custom k_timer to fire in N ticks and have the main thread go idle via _k_cpu_idle()_. If this is workable, then the primary benefit would be that it might allow for the system to sleep for the necessary number of ticks without consuming as much power.

**Comment by cfriedt** [comment]
> The second is ... can we build on this alternate _k_sleep()_ to leverage k_timer in some way? This would require some exploration/experimentation, but the gist of this idea is to set up a custom k_timer to fire in N ticks and have the main thread go idle via _k_cpu_idle()_. If this is workable, then the primary benefit would be that it might allow for the system to sleep for the necessary number of ticks without consuming as much power.

More or less exactly what I was thinking.

I'm less partial about the separate compilation unit, but I do appreciate a reduction in `#ifdef`s occasionally

**Comment by butok** [comment]
Targeting the same issue ;) https://github.com/zephyrproject-rtos/zephyr/pull/66430
Some drivers use k_sleep(). In our case, LPC Serial driver. So compilation of single threaded application is blocked by error. 
We need to move sched.c to common code to unblock our progress.

**Comment by nandojve** [comment]
Hi @peter-mitsis , @cfriedt ,

> The second is ... can we build on this alternate k_sleep() to leverage k_timer in some way? This would require some exploration/experimentation, but the gist of this idea is to set up a custom k_timer to fire in N ticks and have the main thread go idle via k_cpu_idle(). If this is workable, then the primary benefit would be that it might allow for the system to sleep for the necessary number of ticks without consuming as much power.

I think that will be the better idea. However adding a timer inside kernel to track that option could be trick, since we need to manage other resources that may wake-up in the mean time. Because of that, at moment, I added the Peter proposal to the case of a `timeout == K_FOREVER`. This is already handled by the `z_impl_k_sleep` function and system will wait in LPM for an interrupt.

The second part could be introduced in a second moment as an optimization for LPM which is not part of this PR.

File: kernel/sched.c
Comment (andyross) [inline kernel/sched.c:1571]: Are we actually guaranteed to have a timer driver when MULTITHREADING=n?  The original idea was to eliminate everything, including things like timing.  There's no spec for this feature, so... maybe.  But IMHO bloating the system for just this case seems like a bad trade?

You might also consider k_busy_wait(), which is "usually" based on the same system timer, but doesn't have to be.  On Nordic, it's a CPU-counted loop and thus won't require the driver.

**Review by andyross — COMMENTED** [review]
Saw this pop up again.  No complaints, but I do worry that the temptation to add features is cutting against the problems MULTITHREADING=n was designed to solve in the first place.  The original version was an afternoon hack by Ben to allow Zephyr's build system (not so much its kernel) to work to emit a bootloader with as little zephyr code as possible.

File: kernel/sched.c
Comment (nandojve) [reply kernel/sched.c]: Hi @andyross ,

> Are we actually guaranteed to have a timer driver when MULTITHREADING=n? The original idea was to eliminate everything, including things like timing.

I saw that `sys_clock_tick_get_32()` returns 0 when no `CONFIG_SYS_CLOCK_EXISTS`. Otherwise it is well even for `CONFIG_TICKLESS_KERNEL=y`. Besides that I agreed that seems better to use the `k_busy_wait()` since it is there.

**Comment by npitre** [comment]
It might make sense to have a no_multithreading.c that gathers 
all sorts of dummy wrappers and/or simplified implementations. 
Mutex support (PR #63935) is another one that could go in.

**Comment by nandojve** [comment]
Hi @andyross ,

Can you revisit and let me know if something is missing? It important to us have this in v3.6.

**Review by andyross — APPROVED** [review]
As mentioned, no complaints from me, so I'll +1.  I do worry in general that MULTITHREADING=n continues to get thicker, and we don't really have a clear idea about what it's "supposed to do" or which APIs should work in it and how, etc...

So, sure: spinning instead of sleeping is a fine choice.  But what about all the other junk here in sched.c that's probably being pulled into your build, waiting on someone to call it when it doesn't work, etc...

I guess what I'm saying is that we have a fragile design here and someone needs to take this thing over and write up docs for exactly what the feature does and how it works.  For starters: why are we even bothering to build sched.c when literally no scheduling is going to happen?

**Review by cfriedt — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

File: kernel/sched.c
Comment (nashif) [inline kernel/sched.c:1571]: huh, when did we start call those APIs in the code? You should be calling the actual API, not its implementation.

**Pull request merged**
